### ISSUE TITLE
Post-val-pass hang at iter 1 with LoRA rank 16 + 7 target modules on Qwen3-8B-4bit

### ISSUE THREAD

**Issue by mcf-capx** [ml-explore/mlx-lm#1348]
# mlx-lm hangs deterministically after iter-1 val pass on rank-16 + 7-LoRA-module config

## Summary

`mlx_lm.lora` reliably hangs immediately after the iter-1 baseline validation pass completes,
for configurations with LoRA rank ≥ 16 across 7 modules and max_seq_length ≥ 8192. The val
pass log line ("Iter 1: Val loss X.XXX, Val took YYYs") prints successfully, then no further
output is produced and the process consumes <1% CPU indefinitely. A smaller-config "smoke"
training (rank 4, 2 LoRA modules, max_seq 4096) on the same dataset and base model completes
all iterations normally. Reproduced on mlx-lm 0.30.7 and 0.31.3.

## Environment

- Hardware: Apple M5 Max, 128 GB unified memory
- macOS: 26.4.1 (25E253)
- Python: 3.12.12 (homebrew)
- Tested mlx-lm versions: 0.30.7, 0.31.3 (latest at time of writing)
- Base model: `mlx-community/Qwen3-8B-4bit`

## What works (smoke baseline)

```yaml
model: 'mlx-community/Qwen3-8B-4bit'
train: true
data: '<corpus_dir>'
valid_data: '<val.jsonl>'
batch_size: 1
iters: 3
val_batches: 0
steps_per_eval: 99999
steps_per_report: 1
max_seq_length: 4096
grad_checkpoint: true
grad_accumulation_steps: 1
mask_prompt: true
lora_parameters:
  rank: 4
  alpha: 8
  dropout: 0.0
  keys:
    - self_attn.q_proj
    - self_attn.v_proj
learning_rate: 1.0e-4
```

**Result:** iter 1, 2, 3 all complete in ~60s, adapter saves correctly.

## What hangs (production config)

```yaml
model: 'mlx-community/Qwen3-8B-4bit'
train: true
data: '<corpus_dir>'
valid_data: '<val.jsonl>'
batch_size: 1
iters: 1500
val_batches: 25
steps_per_eval: 99999
steps_per_report: 20
max_seq_length: 16384
grad_checkpoint: true
grad_accumulation_steps: 1
mask_prompt: true
lora_parameters:
  rank: 16
  alpha: 32
  dropout: 0.05
  keys:
    - self_attn.q_proj
    - self_attn.k_proj
    - self_attn.v_proj
    - self_attn.o_proj
    - mlp.gate_proj
    - mlp.up_proj
    - mlp.down_proj
learning_rate: 1.0e-4
```

**Result:** val pass completes cleanly with a sensible val_loss, then the process hangs
indefinitely. No subsequent iter is logged. Process consumes <1% CPU and is unresponsive to
SIGTERM (requires SIGKILL).

### Reproducible log signature

```
Trainable parameters: 0.533% (43.647M/8190.735M)
Training
Starting training..., iters: 1500
Calculating loss...:   0%|          | 0/25 [00:00<?, ?it/s]
... (val batches 1-25 progress normally) ...
Calculating loss...: 100%|██████████| 25/25 [03:20<00:00,  8.03s/it]
Iter 1: Val loss 0.858, Val took 200.836s
<-- HANG HERE -->
```

## Diagnostic differential

Bug was tested under every combination of the following without success:

| Variable | Values tried | Effect |
|---|---|---|
| mlx-lm version | 0.30.7, 0.31.3 | both hang identically |
| max_seq_length | 8192, 16384 | both hang (val pass works at both) |
| LoRA rank | 8, 16, 32, 64 | all hang ≥8 with 7 modules |
| LoRA module count | 2, 7 | works at 2, hangs at 7 |
| val_batches | 0, 1, 25 | all hang post-val |
| grad_accumulation_steps | 1, 4 | both hang |
| resume_adapter_file | with/without | both hang |
| Launcher | `python -m mlx_lm lora`, custom wrapper | both hang |
| Full Mac reboot | yes | did not fix |

The unique trigger appears to be **rank ≥ 16 combined with 7 LoRA target modules** at any
sequence length that's nontrivial enough to require non-trivial Metal activations during
the val pass.

## Notes

- The first time we hit this hang, we had completed 240 training iters successfully
  (iter 1 val passed, then 240 train iters at ~30 sec/iter, then iter 250 val hung).
  Every subsequent invocation of this same config — even after killing all processes,
  rebooting the Mac, and deleting checkpoint files — hangs at the FIRST val pass.
  This suggests the hang is not caused by accumulated Metal state across iterations,
  but the trigger does seem to be the val pass itself in this config shape.
- `caffeinate` and `osascript` notification calls were active during training (run via
  a parent shell). Doesn't seem related — confirmed with bare invocation too.
- The `mx.metal.device_info is deprecated` warning is emitted on 0.30.7 but not 0.31.3.
- Smoke config and failing config use the IDENTICAL corpus directory and validation set —
  data is not the variable.

## Minimal repro

The failing config is reproducible on any sufficiently long-input corpus with the
Qwen3-8B-4bit base. Suggested synthetic repro (untested, but should preserve the
input-length characteristics):

```python
# tools/generate_synth_long_corpus.py
import json
LOREM = "Lorem ipsum dolor sit amet consectetur adipiscing elit. " * 100  # ~5500 tokens
rows = []
for i in range(100):
    rows.append({
        "messages": [
            {"role": "system", "content": LOREM * 2},  # ~11k tokens
            {"role": "user", "content": f"Question {i}: " + LOREM[:300]},
            {"role": "assistant", "content": json.dumps({"answer": f"Response {i}"})},
        ]
    })
with open("train.jsonl", "w") as f:
    for r in rows: f.write(json.dumps(r) + "\n")
```

## Expected behavior

The val pass should complete, the training loop should proceed to iter 1 forward+backward,
and `Iter 20: Train loss X.XXX` should appear in the log (per `steps_per_report: 20`).

## Workarounds attempted (all failed)

1. Reduce `max_seq_length` to 8192 — same hang
2. Reduce `val_batches` to 0 — same hang
3. Set `grad_accumulation_steps: 1` — same hang
4. Try `lora_cached_prefix.py` wrapper script (custom local wrapper) — same hang
5. Downgrade to mlx-lm 0.30.7 — same hang
6. Full Mac reboot — same hang
7. Delete adapter dir and start fresh — same hang

**Comment by mcf-capx** [comment]
Standalone repro script (one command, no proprietary data) — synthesizes a Lorem Ipsum corpus and triggers the hang reproducibly: https://gist.github.com/mcf-capx/848214a243974c463ecdb7f21a563879

To run:

```
curl -sSL <RAW_GIST_URL>/raw/reproduce_mlx_lm_hang.py | python3
```

Expected (buggy): val pass logs cleanly, then no further output for >5 min, process at <1% CPU. Expected (fixed): `Iter 20: Train loss X.XXX` appears within ~5 min.
